model/Mastermind.php

```diff
@@ -183,4 +216,50 @@ public function decrementRemainingAttempts(): void
         $this->remainingAttempts--;
     }
 
+    /**
+     * @param Combination $proposition
+     * @throws Exception Si la partie n'existe pas ou qu'elle est terminée
+     */
+    public function play(Combination $proposition): void
+    {
+        if ($this->remainingAttempts <= 0) {
+            throw new Exception('Partie non créée');
+        } else {
+            $this->addProposition($proposition);
+            $comparator = new CombiComparator($this->getSolution());
+            $this->addCompareResult($comparator->compare($proposition));
+            $this->decrementRemainingAttempts();
+        }
+    }
+
+    /**
+     * Renvoie true si la partie n'est ni gagné ni perdu
+     * @return bool
+     */
+    public function isGameOver(): bool
+    {
+        return ($this->isGameLost() || $this->isGameWon());
+    }
+
+    /**
+     * Renvoie true si la partie est perdue
+     * @return bool
+     */
+    public function isGameLost(): bool
+    {
+        return $this->remainingAttempts <= 0;
+    }
+
+    /**
+     * Renvoie true si la partie est gagnée
+     * @return bool
+     */
+    public function isGameWon(): bool
+    {
+        if (isset($this->compareResults[count($this->compareResults) - 1])) {
+            return $this->compareResults[count($this->compareResults) - 1]->getBlackPaws() === $this->getSize();
+        }
+        return false;
```

Comment:
Je pense que le plus simple est de comparer la dernière proposition à la solution. Il me semble que tu peux utiliser l'operateur `==` et `===` pour comparer des tableaux. si $a et $b sont des tableaux alors : 
- `$a == $b` est vrai si ils ont les mêmes clés et les même valeurs
- `$a === $b` est vrai si ils ont les mêmes clés et les même valeurs dans le même ordre.

Je te laisse tester ça ;)
([cf php.net](https://www.php.net/manual/en/language.operators.array.php))

Reply:
J'ai testé juste comme ça dans le fonction isGameWon, pour voir :
$proposition = $this->propositions[count($this->propositions) - 1];
        $combination = $proposition->getPaws();
        $solution = $this->solution->getPaws();
        if ($combination == $solution) {
            $result = true;
        } else {
            $result = false;
        }
        var_dump($result);
Et cela marche bien avec ==
Par contre cela ne marche pas avec ===

Reply:
Du coup, cela donnerait ça pour la fonction isGameWon :
public function isGameWon(): bool
    {
        if (isset($this->propositions[count($this->propositions) - 1])) {
            $combination = $this->propositions[count($this->propositions) - 1];
            $proposition = $combination->getPaws();
            $solution = $this->solution->getPaws();
            return $proposition == $solution;
        }
        return false;
    }

Reply:
Et en un peu plus cours :
public function isGameWon(): bool
    {
        if (isset($this->propositions[count($this->propositions) - 1])) {
            $proposition = $this->propositions[count($this->propositions) - 1];
            return $proposition->getPaws() == $this->solution->getPaws();
        }
        return false;
    }

Reply:
```php
$count = count($this->propositions);
return $count > 0 ? $this->propositions[$count] == $this->solution->getPaws();
```

Reply:
Il manque quelques choses non ? 
Ce serait ça plutôt non ? return $nbProposition > 0 ? $this->propositions[$nbProposition - 1] == $this->solution->getPaws() : false;
Mais cela ne fonctionne pas, on ne gagne jamais la partie.
Mon isset au début était là car la première fois, la valeur de la première ligne est null puis qu'il n'y a pas encore de proposition.

Reply:
C'est mieux comme-ça : 
return $nbProposition > 0 ? ($this->propositions[$nbProposition - 1]->getPaws() == $this->solution->getPaws()) : false;
;)

Reply:
En effet, t'étais fatigué hier. ;)

Reply:
oui il manque le `: false`effectivement.

> Mon isset au début était là car la première fois, la valeur de la première ligne est null puis qu'il n'y a pas encore de proposition.

normalement `$nbProposition > 0 ? `  est équivalent à `if (isset($this->propositions[count($this->propositions) - 1]))` : si nbProposition == 0 alors il n'y a pas de proposition.

Essaye peut être d'installer l'addon firefox [xdebug helper](https://addons.mozilla.org/fr/firefox/addon/xdebug-helper-for-firefox/) puis tu actives le debug comme sur l'image :
![xdebug helper](https://user-images.githubusercontent.com/4210719/99780010-2e3da980-2b16-11eb-8b96-a3d89af805bd.png)
Ensuite tu met un point d'arrêt dans ton code pour voir si ça marche

Reply:
Le point d'arrêt, je le met depuis où ? Si dans Firefox, comment ?

Reply:
Vu que cela fonctionne déjà, c'est juste pour infos, pour plus tard, car ça pourrait servir.

Reply:
- Le point d'arrêt tu le mets dans ton code sur phpStorm : 

![breakpoint](https://user-images.githubusercontent.com/4210719/99785551-9643be00-2b1d-11eb-95d1-fad93447628c.png)

- pense à écouter les connexions de debug (icône de téléphone en haut à droite) : 

![php debug](https://user-images.githubusercontent.com/4210719/99785754-e02ca400-2b1d-11eb-8073-408952f301d5.png)

- si le debug fonctionne et que l'execution du code s'arrete au niveau du point d'arrêt, tu peux sélectionner la ligne `$nbProposition > 0 ? ($this->propositions[$nbProposition - 1]->getPaws() == $this->solution->getPaws()) : false;` (sans le return) et appuyer sur [alt]-[F8] puis [Entrée] pour simuler l'execution de la ligne sélectionnée.

Reply:
Je crois que ça fonctionne mais c'est pas clair du tout pour l'utilisation dans PHPStorm.
Faudra qu'on regarde un peu ça ensemble quand t'auras 15 minutes après ton taf ou dans le week-end.
En attendant, je ferai sans ça. ;)